### PR TITLE
Validation for Vulkan SC 1.0.11 (based on 1.2.198)

### PR DESCRIPTION
**Pull request by dgkoch**
Commands for building the VulkanSC version of the layers are similar to the instructions for Vulkan with a few minor differences:

- When running "update_deps.py" for your platform, the option "--api vulkansc" should be specified.
- When running "cmake -C helper.cmake ..." the option "-DVULKANSC=ON" should be passed in.

### CODE REVIEW

**Review by jeremyg-lunarg — CHANGES_REQUESTED** [review]
Several general points:

- this PR tries to merge to the sdk_1.2.198 branch, which is not correct as that would destroy the ability to recreate that SDK
- the number of ifdefs added is going to create a huge maintenance burden.   I'm especially concerned about the extension ifdefs as it will be very unclear to non-SC developers which extensions need to be ifdefed. Is it extensions that are not allowed by SC, or extensions that are in a newer Vulkan-Headers version that SC isn't using yet? Developers are going to forget to ifdef things and have to repush changes frequently to fix ifdef related build breaks.
- The generated code being duplicated is not ideal.  My general impression is that SC is mostly a subset of Vulkan. It only matters if the application uses stuff outside the stuff in the subset, VVL could still have generated code for 'everything'. 

It would be MUCH better to make a SC specific runtime mode for VVL. We have magic SC bits in the version number that could enable it. It is already possible to validate a Vulkan 1.0 app using the very latest VVL code built with 1.3.xx and support for all extensions.   It seems like a better approach would be to add a new SC ValidationObject (similar to Best Practices or Sync Validation) whose only job is to detect accesses to functionality forbidden by SC.

Doing this probably raises requirements on Vulkan-Headers, especially vk.xml.   Ideally we'd always compile validation using vulkan.h and be able to tell what is allowed in SC by looking at fields in vk.xml.   (having SC applications compile against vulkan_sc.h does make sense as it provides them an immediate way to detect when they're using something they shouldn't) It looks like there's a separate sc_main branch of Vulkan-Headers, is that how things are going to be in the long term? If Vulkan-Headers is going to merge main and sc_main, do you have a plan for how that would work?

**Review by ncesario-lunarg — CHANGES_REQUESTED** [review]
Basically echoing @jeremyg-lunarg's comments, especially WRT the "SC specific runtime mode," as that could make this sort of change tenable.

Aside from that, regarding the current changes, the ~1000 pre-processor "ifdefs" effectively split the current code base into a "Vulkan" and "Vulkan SC" codebase, and are untenable as is. This split appears to be relatively logical for the generated code (separate files/directory), but a similar split would need to be made for the rest of the code as well. I could see one approach would be to copy the existing validation code to a sub directory with the ifdef'd code stripped out, though there are arguably cleaner/clearer ways to do this.

File: layers/CMakeLists.txt [hunk 1]
```diff
@@ -138,11 +151,37 @@ if(CMAKE_C_COMPILER_ID MATCHES "Clang")
...
+# and xxhash.* are included to avoid unresolved symbol errors
+# at runtime.
+set(CHASSIS_LIBRARY_FILES
+    ${GENERATED_DIR}/chassis.cpp
+    ${GENERATED_DIR}/layer_chassis_dispatch.cpp
+    ${GENERATED_DIR}/vk_safe_struct.cpp
+    ${GENERATED_DIR}/vk_safe_struct.h
+    layer_options.cpp
+    xxhash.h
+    xxhash.c
```
[the remark this answers is not in the document]
Comment (dgkoch) [reply layers/CMakeLists.txt:167]: they just aren't working yet.

File: layers/best_practices_validation.h [hunk 2]
```diff
@@ -408,21 +431,27 @@ class BestPractices : public ValidationStateTracker {
     bool PreCallValidateGetPhysicalDeviceQueueFamilyProperties2(VkPhysicalDevice physicalDevice,
                                                                 uint32_t* pQueueFamilyPropertyCount,
                                                                 VkQueueFamilyProperties2* pQueueFamilyProperties) const override;
+#if defined(VK_KHR_get_physical_device_properties2)
```
[the remark this answers is not in the document]
Comment (dgkoch) [reply layers/best_practices_validation.h:434]: VKSC did not retain the extension form of of extension functionality that was promoted to core VK 1.2 or earlier, only the core tokens and entry points exist.

**Comment by dgkoch** [comment]
> this PR tries to merge to the sdk_1.2.198 branch, which is not correct as that would destroy the ability to recreate that SDK

The goal was never to merge to this branch - I'd requested that an `sc_main` branch be created based on this tag, but that request was refused. 

> the number of ifdefs added is going to create a huge maintenance burden. Developers are going to forget to ifdef things and have to repush changes frequently to fix ifdef related build breaks.

I understand this is a concern - on the other hand I'm not sure a runtime mode make sense either, as there's probably never going to be a desire to switch VVL between VK and VKSC without re-compiling...  

>  I'm especially concerned about the extension ifdefs as it will be very unclear to non-SC developers which extensions need to be ifdefed. Is it extensions that are not allowed by SC, or extensions that are in a newer Vulkan-Headers version that SC isn't using yet?

The extensions which ifdefed are because (a) we didn't keep extension form of functionality that was promoted to core in VK 1.2 or earlier (VKSC 1.0 is based on VK 1.2) or (b) the extensions aren't present in the VKSC header (either we chose not to expose them for VKSC, or they are newer than the current VKSC spec branch).  We could have made it ifdef (VULKAN_SC), but it seemed to make more sense to predicate it on the actual extension strings that added the functionality. 

>  Developers are going to forget to ifdef things and have to repush changes frequently to fix ifdef related build breaks.

I mean - that's what CI is for..

We'll have to think more about how these can co-exist with less impact.

File: layers/CMakeLists.txt [hunk 3]
```diff
@@ -23,7 +34,9 @@ if(WIN32)
 elseif(ANDROID)
     add_definitions(-DVK_USE_PLATFORM_ANDROID_KHR -DVK_USE_PLATFORM_ANDROID_KHX)
 elseif(APPLE)
-    add_definitions(-DVK_USE_PLATFORM_MACOS_MVK -DVK_USE_PLATFORM_METAL_EXT)
+    if(NOT VULKANSC)
+        add_definitions(-DVK_USE_PLATFORM_MACOS_MVK -DVK_USE_PLATFORM_METAL_EXT)
```
[the remark this answers is not in the document]
Comment (dgkoch) [reply layers/CMakeLists.txt:38]: Likely because the build was only tested on Linux and Mac. 
The VKSC headers don't include the macos headers.

**Comment by KarenGhavam-lunarG** [comment]
>> this PR tries to merge to the sdk_1.2.198 branch, which is not correct as that would destroy the ability to recreate that SDK

> The goal was never to merge to this branch - I'd requested that an sc_main branch be created based on this tag, but that request was refused.

Yes, an sc_main branch was refused because it was premature to assume the Vulkan safety critical code would eventually be merged into this code base (that an sc_main branch would be suggesting). This PR's approach of a compile time solution with many ifdef's was determined to not be maintainable with an undesirable impact on the broader Vulkan ecosystem.

> We'll have to think more about how these can co-exist with less impact.

Agreed.

**Comment by jeremyg-lunarg** [comment]
> I understand this is a concern - on the other hand I'm not sure a runtime mode make sense either, as there's probably never going to be a desire to switch VVL between VK and VKSC without re-compiling...
> 
> >

There isn't a desire to runtime switch between Vulkan and SC for end users of validation. It does make this vastly simpler for us to maintain and probably will reduce the CI burden.

Well over 95% of the developers working on validation and the end users of validation will be using Vulkan, and not SC.  Any coexistence solution should acknowledge that.  Ifdefing every single Vulkan extension because SC doesn't have them has the burden reversed, so that normal Vulkan development has to bend over backwards to make sure to not break SC.  Whatever solution we end up with should try to make SC 'draft' off of normal Vulkan development without interfering.  Having a runtime mode is my best guess at how we could accomplish this, but if there are other ways to do it I'd be open to discussing them.

**Comment by jeremyg-lunarg** [comment]
Closing this, because SC validation is going to be further developed in a fork until it is easier to determine if coexistence in the same code base is feasible.

**Pull request closed without merging**